source/opt/constructs.h

```diff
@@ -0,0 +1,229 @@
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// This file defines the language constructs for representing a SPIR-V
+// module in memory.
+
+#ifndef LIBSPIRV_OPT_CONSTRUCTS_H_
+#define LIBSPIRV_OPT_CONSTRUCTS_H_
+
+#include <functional>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "spirv/spirv.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+class Function;
+class Module;
+
+// Operand is a overloaded term. For better understanding, in this file, we
+// use payload to mean what "operand" is supposed to mean in the SPIR-V spec.
+// Operand itself is reserved to mean the real operands to an instruction,
+// excluding type ids and result ids.
+
+// A payload to an Inst. It can be a type id, a result id, or a real logical
+// operand.
+struct Payload {
+  Payload(spv_operand_type_t t, std::vector<uint32_t>&& w)
+      : type(t), words(std::move(w)) {}
+
+  spv_operand_type_t type;      // Type of this payload.
+  std::vector<uint32_t> words;  // Binary segments of this payload.
+
+  // TODO(antiagainst): create fields for literal number kind, width, etc.
+};
+
+class Inst {
+ public:
+  Inst() : opcode_(SpvOpNop), type_id_(0), result_id_(0) {}
+  Inst(SpvOp op) : opcode_(op), type_id_(0), result_id_(0) {}
+  Inst(const spv_parsed_instruction_t& inst, std::vector<Inst>&& dbg_line = {});
+
+  SpvOp opcode() const { return opcode_; }
+  uint32_t type_id() const { return type_id_; }
+  uint32_t result_id() const { return result_id_; }
+  // Returns the first word of |index|th operand.
+  uint32_t GetSingleWordOperand(uint32_t index) const {
+    return GetSingleWordPayload(index + TypeResultIdCount());
+  }
+  uint32_t NumOperandWords() const;
+  // Returns the payload at the |index|th operand.
+  const Payload& GetOperand(uint32_t index) const {
+    return GetPayload(index + TypeResultIdCount());
+  }
+  uint32_t NumOperands() const {
+    return static_cast<uint32_t>(payloads_.size() - TypeResultIdCount());
+  }
+  // Returns the first word of |index|th payload.
+  uint32_t GetSingleWordPayload(uint32_t index) const;
+  uint32_t NumPayloadWords() const {
+    return NumOperandWords() + TypeResultIdCount();
+  }
+  // Returns the |index|th payload.
+  const Payload& GetPayload(uint32_t index) const;
+  uint32_t NumPayloads() const {
+    return static_cast<uint32_t>(payloads_.size());
+  }
+  void SetPayload(uint32_t index, std::vector<uint32_t>&& data);
+  std::vector<Inst>& dbg_line_info() { return dbg_line_info_; }
+
+  bool IsNop() const {
+    return opcode_ == SpvOpNop && type_id_ == 0 && result_id_ == 0 &&
+           payloads_.empty();
+  }
+  void ToNop() {
```

Comment:
We'll need this for transformations indicating "discarding" this instruction.

No, I don't think "debug_line_info" should be cleared at the same time since those line-related debug instructions are just placed here. They are not a part of this instruction.